### PR TITLE
Add remind me later support to task list

### LINKED ISSUE #6637: Dismiss and Remind me later in task ellipses menu
This is an update to the task list components on the home screen - Store setup and Extended task list.

- There's currently a `Dismiss` action that appears as a full word on the right side of a dismissible task. This update would move that action to an ellipses menu.
- The `Remind me later` action would set a count of 24 hours for the task to appear again in the list.
- The only tasks in the store setup task list that are dismissible are the paid extensions install tasks

Tracks events:
`wcadmin_extended_tasklist_task_click`
property: `task_name`
When user selects task
`wcadmin_extended_tasklist_task_dismiss`
property: `task_name`
When user dismisses task from ellipses menu
`wcadmin_extended_tasklist_task_remindmelater` 
property: `task_name`
When user selects "Remind me later" from ellipses menu

### PR DESCRIPTION
**Pull request by louwie17**
Fixes #6637 

This replaces the Dismiss button in the task list item to an ellipsis dropdown menu, this contains the dismiss button and also a 'remind me later' option if that option is enabled.
To align the ellipsis menu properly with the header, I switched the task layout to css grid. 

### Accessibility

-   [x] I've tested using only a keyboard (no mouse)
-   [ ] I've tested using a screen reader
-   [x] All animations respect [`prefers-reduced-motion`](https://developer.mozilla.org/en-US/docs/Web/CSS/@media/prefers-reduced-motion)
-   [x] All text has [at least a 4.5 color contrast with its background](https://webaim.org/resources/contrastchecker/)

### Screenshots

remind-me-later

### Detailed test instructions:

-  Install the task list tester plugin from here (https://github.com/louwie17/woocommerce-admin-tasklist-helper/releases/tag/v1.0.0)
- Make sure the above plugin is enabled.
- Navigate to **WooCommerce > Home** having finished the onboarding flow
- The extended task list **Things to do next** should be present.
- Click on the three dot ellipsis menu for the item `Set up the task list tester` task 
- The dropdown should include `Remind me later`, click on this
- The task should disappear and trigger the `wcadmin_extended_tasklist_remindmelater_task` event, it should also stay gone on refresh
- This should dismiss it for 24hours, but we can shorten this by updating the `woocommerce_task_list_remind_me_later_tasks` option.
- Using the [WCA test helper](https://github.com/woocommerce/woocommerce-admin-test-helper/releases/tag/v0.3.0) go to **Tools > WCA Test Helper > Options** and search for the above option
- Edit the option and change the timestamp (in milliseconds) of the `woocommerce-admin-tasklist-tester-setup` key to something more recent (you can use [epochconverter](https://www.epochconverter.com/) just append `000` at the back for milliseconds)
- After the set wait timestamp has passed and you refresh the page, the task list item should appear again.
- Try the `Remind me later` again, and click the `undo` notice, it should make the task show up again.

### CODE REVIEW

File: client/task-list/task-item.tsx
Comment (samueljseay) [inline client/task-list/task-item.tsx:137]: Is this type really correct?

File: client/task-list/task-list.js
Comment (samueljseay) [inline client/task-list/task-list.js:216]: I think this is complex enough that it deserves either a integration test (ie a test of the component's displaying of remindMeLaterTasks) or it needs to be refactored to take some of these things as args and unit tested. Thats just my gut feeling. This component has a lot of complexity in it.

File: packages/components/src/ellipsis-menu/test/index.js
Comment (samueljseay) [inline packages/components/src/ellipsis-menu/test/index.js:11]: Thanks so much for adding some tests for this component!

File: packages/components/src/ellipsis-menu/index.js
Comment (samueljseay) [inline packages/components/src/ellipsis-menu/index.js:78]: Gosh, I wish we could change this to children 😢 

Not gonna demand a change here, but I'd be grateful if we supported children in a backward compatible way like we originally considered with the List component.

File: packages/components/src/ellipsis-menu/test/__snapshots__/index.js.snap
Comment (samueljseay) [inline packages/components/src/ellipsis-menu/test/__snapshots__/index.js.snap:1]: Whats your thought on snapshot tests? I don't see them as being high value, but don't mind keeping them if you see the value.

File: packages/components/src/ellipsis-menu/test/index.js
Comment (samueljseay) [inline packages/components/src/ellipsis-menu/test/index.js:4]: Worth mentioning that testing library recommend using the higher level lib `user-event` for a more realistic way to fire events as the user. https://github.com/testing-library/user-event We already have this available as a dependency. not a huge deal in this case as its not used widely in these tests, but can't hurt to form a single approach.

**Review by samueljseay — COMMENTED** [review]
This is looking awesome. Just a range of minor suggestions from me. Let me know what you think.

File: client/task-list/task-item.tsx
Comment (louwie17) [reply client/task-list/task-item.tsx:137]: I will stick with `MouseEvent`.

File: packages/components/src/ellipsis-menu/index.js
Comment (louwie17) [reply packages/components/src/ellipsis-menu/index.js:78]: This one is actually more justifiable as the `Dropdown` component it uses, uses the exact same format. It would be nice if the `Dropdown` component was converted first to make use of children.

File: client/task-list/task-list.js
Comment (louwie17) [reply client/task-list/task-list.js:216]: I added some extra test cases in this commit, specifically targeting the `getVisibleTasks` method - https://github.com/woocommerce/woocommerce-admin/pull/6923/commits/c44d79ac84a0400d13a7168c3ae7fdcc79b0ef42

I do agree that this component has a lot of complexity (too much I would say), some of the future work Jeff has been poking at, should help with extracting this I think.

**Comment by louwie17** [comment]
@samueljseay it should be good for a re-review?

File: packages/components/src/ellipsis-menu/index.js
Comment (samueljseay) [reply packages/components/src/ellipsis-menu/index.js:78]: That makes sense 🙂

**Comment by samueljseay** [comment]
One thing that would really help to review the functionality is an explanation of the logic behind what tasks support "remind me later". I have asked a question on the original issue because this is a cause of confusion for me. Is it supposed to be only dismissable tasks that have a "remind me later" or should all tasks?

**Comment by louwie17** [comment]
@samueljseay is this good to go now from Elizabeth's response here: https://github.com/woocommerce/woocommerce-admin/issues/6637#issuecomment-837075677

**Comment by samueljseay** [comment]
@louwie17 still some confusion from me on this. If its for all tasks then why is dismiss support not added to the main task list as part of this? the issue specifies it should support on dismiss for "Store setup and extended task list". Is the intent to do that in a follow up PR? If it is I guess we should make it clear this doesn't close #6637 ?

Forgive me if I'm missing something here. 🙇

**Comment by becdetat** [comment]
IMO this is behaviour that should be tested by GS before release, as this will be relied on by 3pds and internally in a8c in other projects. Given that it can only be tested by adding a new task could you create a little plugin that GS could download and install, rather than having to create a plugin from scratch (which GS can't do)?

File: client/task-list/task-item.tsx
Comment (becdetat) [inline client/task-list/task-item.tsx:120]: This logic is a bit complicated, could you move it up into a `showEllipsisMenu` const?

**Comment by louwie17** [comment]
> IMO this is behaviour that should be tested by GS before release, as this will be relied on by 3pds and internally in a8c in other projects. Given that it can only be tested by adding a new task could you create a little plugin that GS could download and install, rather than having to create a plugin from scratch (which GS can't do)?

Great suggestion @becdetat, I created a [task list tester plugin](https://github.com/louwie17/woocommerce-admin-tasklist-helper/releases/tag/v1.0.0) and updated the above testing instructions, I will have to add testing instructions to this before merging.

**Comment by jeffstieler** [comment]
What do you think about moving the code from your testing plugin into our "add task" example?

I'll be reviewing the code shortly, here are some things I noticed in testing:

* The animations seems a little busy when postponing a task
Kapture 2021-05-14 at 11 37 52
* The notices don't auto-dismiss, or dismiss on click
* Do we want to track the "undo" click? I didn't see an event fire for that

**Comment by louwie17** [comment]
> What do you think about moving the code from your testing plugin into our "add task" example?

Yeah that would probably also be helpful, although would still need the other plugin for GS testing.

> I'll be reviewing the code shortly, here are some things I noticed in testing:

Thanks :) 

> * The animations seems a little busy when postponing a task

Yeah that does look kinda odd, a fade in/out might look nicer, @elizaan36 would you have any thoughts on this?
There is technically 3 things happening (the dismissed item disappearing, a item previously hidden appearing, and the show/hide bar disappearing). 

> * The notices don't auto-dismiss, or dismiss on click

I will have a look, I thought I use the same as the the current `dismiss` notice.

> * Do we want to track the "undo" click? I didn't see an event fire for that

We might, @pmcpinto do you have a preference?

**Comment by elizaan36** [comment]
Thanks for the ping!

> Yeah that does look kinda odd, a fade in/out might look nicer, @elizaan36 would you have any thoughts on this?
There is technically 3 things happening (the dismissed item disappearing, a item previously hidden appearing, and the show/hide bar disappearing).

If the task was dismissed in-line vertically, and the new task appeared vertically rather than sliding off to the right or in to the left it would make sense. I think I'm reacting mainly to the movement right and left.

The footer could animate up to disappear as well? I was thinking something more like this, where the task and footer animate up as they disappear and the list collapses vertically to close the gap:
Screen Recording 2021-05-17 at 04 11 43 PM

**Comment by louwie17** [comment]
Thanks @elizaan36 I will update that and see how it looks.

**Comment by louwie17** [comment]
> If the task was dismissed in-line vertically, and the new task appeared vertically rather than sliding off to the right or in to the left it would make sense. I think I'm reacting mainly to the movement right and left.
>
> The footer could animate up to disappear as well? I was thinking something more like this, where the task and footer animate up as they disappear and the list collapses vertically to close the gap:

@elizaan36 @jeffstieler So after messing around with the transitions for probably too long I landed on this, which I am still not happy with, as it still feels like a lot is happening. The footer collapsing part is fine, it's mostly the transitioning in/out of the dismissed task item (more context below).

task-list-animations

I ran into some weird issues with the transition library we use in particular when we are removing and adding an item at the same time.
(This below might be too much context)
Mostly because when we dismiss the last item (Example) the new item currently hidden will show up above Example. This causes a quick jump, and makes the moving up (out of the list) transition not quite right, this does seem to work fine when dismissing the top item as it stays at the top when transitioning out. This is more clearly visible in the gif Jeff posted above.

Doing a in/out fade of the task list item did seem to look pretty good, but would still require some logic altering depending on which item is being dismissed see gif below (slowed down for demo purposes):

task-list-transition-fade

All in all, this could be fixed by altering the logic, but that would take some extra time, as it is a little confusing. Although removing the use of the transition library and doing it manually could be easier (I haven't tried that yet).
@elizaan36 @jeffstieler would you have a potential alternate transition suggestion? Or stick with one of the above (altering the logic a bit)?
Lastly maybe someone else on @woocommerce/ghidorah who is good with transitions wanted to have a look :) (I might of stared at the code for toooo long)

**Comment by louwie17** [comment]
@octaedro suggested staggering the transitions, so it doesn't do two things at the same time, which I might try to see how that looks.

**Comment by elizaan36** [comment]
Staggering the transitions is a good idea between the dismissed task and the footer. 

There's also the issue of the dismissed task fading "above" the other tasks. The dismissed task shouldn't appear over any other content on the screen.

Not sure if this requires a tweak to the transition library but could you try shifting the position of the element as it's dismissed so that a dismissed task only ever fades out "below" the remaining tasks? That would create more of a subtle transition.

**Comment by louwie17** [comment]
@jeffstieler this PR should be ready for re-review if you feel up for it :) I addressed your points:

> The animations seems a little busy when postponing a task

Those are updated now and got approval from Elizabeth:
task-list-animations

> The notices don't auto-dismiss, or dismiss on click

This was a separate issue and was in a separate PR (this should be working now as I rebased to the latest)

> Do we want to track the "undo" click? I didn't see an event fire for that

I added an event for this incase we do want that data - `tasklist_undo_remindmelater_task`

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (jeffstieler) [inline packages/experimental/src/experimental-list/collapsible-list/index.tsx:127]: Is this required? It seems like it won't change from the initial values.

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (jeffstieler) [inline packages/experimental/src/experimental-list/collapsible-list/index.tsx:156]: A comment explaining how this staggers the animation would be helpful.

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (jeffstieler) [inline packages/experimental/src/experimental-list/collapsible-list/index.tsx:60]: Is this covered by tests?

**Review by jeffstieler — APPROVED** [review]
The code here looks good and tested well - nice work on the staggered animations. I've got a question about the footer text being in state, but this looks good to merge (feedback optional).

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (louwie17) [reply packages/experimental/src/experimental-list/collapsible-list/index.tsx:127]: Not overly, I did it because of the staggered transition. Because during the transition you would see the label change to `Show 0 items`, which looked a little odd. This way it keeps the same label when it minimizes.
I could push that logic upwards otherwise.

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (louwie17) [reply packages/experimental/src/experimental-list/collapsible-list/index.tsx:60]: No, I tried writing some tests of the actual staggered transition, but jest didn't like the `useFakeTimers` very much.
I will add tests for this function.

File: packages/experimental/src/experimental-list/collapsible-list/index.tsx
Comment (jeffstieler) [reply packages/experimental/src/experimental-list/collapsible-list/index.tsx:127]: A comment in the code would suffice here - thanks for the explanation.

**Pull request merged**
